### DISCUSSION TITLE
iCub redball demo iCubZagreb01

### DISCUSSION THREAD

**Discussion by ivan9a** [robotology/community#553] [Q&A] [answered]
Hi,

Do you have some instruction how to properly adjust demo program redball. If we run redball demo we have a significant error between hand and the grasping object. I think that this error is not caused by wrong camera parameters, because I always check parameters before I run pf3Dtracker and demoRedBall. Please take a look at pictures in the attachment:

![thumbnail_IMG_20210929_095040](https://user-images.githubusercontent.com/83449933/135227682-fe823191-8dad-42ab-9d91-f5fb073adf18.jpg)
![thumbnail_IMG_20210929_095106](https://user-images.githubusercontent.com/83449933/135227690-22610de8-2b97-41fc-bf09-191ad8340db9.jpg)
![thumbnail_IMG_20210929_095136](https://user-images.githubusercontent.com/83449933/135227692-9f9d77bc-69d9-4e6f-82df-73f0a1a62c23.jpg)
![thumbnail_IMG_20210929_0950407](https://user-images.githubusercontent.com/83449933/135227695-abed57e1-cff8-4d22-8944-e38249864e04.jpg)
![thumbnail_IMG_20210929_0950550](https://user-images.githubusercontent.com/83449933/135227701-234f2a1d-743a-4409-aca2-35320b1cc51d.jpg)

Do you have a program that would run in parallel with the redball demo program and that would allow simple conversation with the robot. 

Thank you,
Best Regards,
Ivan Župančić

**Comment by andreaderito** [comment]
Dear @ivan9a ,
thank you, let me cc a few colleagues of ours to follow up this issue.
cc @Uboldi80 @pattacini @vtikha @maggia80

**Comment by vtikha** [comment]
Dear @ivan9a, 

I am sorry to see that you are still having some issues with fine-tuning the robot to use the `redball` demonstration. If you look at our last conversation on the [8th of June](https://github.com/robotology/community/discussions/513#discussioncomment-841622), we provided details on how to adjust these offsets, as mentioned:

> Regarding the error you observed, the demo itself provides reaching offsets to adjust the relative position between the ball and the hands. You can either adjust [such offsets manually](https://github.com/robotology/icub-basic-demos/blob/ba9a10343635741b464abe04cbadb57d2f55eff4/demoRedBall/app/conf/config.ini#L22) or rely on the [Grasp the Ball Application](https://robot-bazaar.iit.it/apps/academy/courses/1) which does the procedure automatically if specified. In the latter case, you can follow the instructions at the provided link for downloading and starting the application.

In case you want to do it manually, you need to change the **reach** and **grasp** offset of each arm 

```
reach_offset               0.0 -0.05 0.0 
grasp_offset               0.0  0.00 0.0
```

These values are expressed in `meters` using the `cartesian space` of the robot as reported on the image below (Clicking on the image will take you to the official documentation page.): 

 [<img src="https://icub-tech-iit.github.io/documentation/icub_kinematics/icub-forward-kinematics/img/img-3.jpg">](https://icub-tech-iit.github.io/documentation/icub_kinematics/icub-forward-kinematics/icub-forward-kinematics/) 

As for the `conversational` part of the question, we also replied to this on the link above and you can `pre-script` the robot to say things at particular instances, or endeavour in applications using `google cloud services` as mentioned:

>- regarding the **speech**, we have [Google speech-to-text](https://robot-bazaar.iit.it/apps/academy/courses/15), [Google Speech Processing](https://robot-bazaar.iit.it/apps/academy/courses/19), and [Google text-to-speech](https://robot-bazaar.iit.it/apps/academy/courses/17), which respectively allow you to convert speech to text, to retrieve the sentence's structure and syntax along with the sentiment, and to generate spoken sentences from written text. You can also find the [Google Speech Application](https://robot-bazaar.iit.it/apps/academy/courses/20) that puts the previous three together. Finally, you can find the [Google Dialog](https://robot-bazaar.iit.it/apps/academy/courses/16) application, that allows you to carry on a conversation about iCub world.

**Reply by ivan9a** [reply]
Hi,

Thank you for the clear explanation, but I have already tried to adjust the parameters in the config.ini file. I made changes to x, y, z reaching offset, and ball diameter. The change did not result in any improvement in grasping the ball. After each change, I compiled the redball demo program, but it didn't help either. Please look at the picture below. I think it would be best to arrange a conference call tomorrow afternoon if that suits you.

![Screenshot from 2021-10-03 11-41-54](https://user-images.githubusercontent.com/83449933/135748673-684a340b-f857-47d4-b05b-cce38f824e9d.png)

Thank you,
Best regards,
Ivan Župančić

**Comment by vtikha** [comment, marked as answer]
Dear @ivan9a,

it might be that the `demoRedBall` is not initialised with the correct file. One way check which file is actually being used please run the following command to get the full path. 

`yarp resource --context demoRedBall --from config.ini`

If you are using the ball that we provided with the robot, the `radius` should not be changed. 

Let us know if you manage with the above command.
As for the conference call we we had sent you an email, you can directly reply there.
